bin/git-theta

```diff
@@ -43,45 +51,53 @@ def parse_args():
     return args
 
 
-def paths_and_parameters(model_dict, root):
-    """Convert the keys of the model dict to parameters based starting at root."""
-    return {os.path.join(root, *k): v for k, v in utils.flatten(model_dict).items()}
+def post_commit(args):
+    repo = git_utils.get_git_repo()
+    theta_commits = theta.ThetaCommits(repo)
+
+    gitattributes_file = git_utils.get_gitattributes_file(repo)
+    patterns = git_utils.get_gitattributes_tracked_patterns(gitattributes_file)
+
+    oids = set()
+    commit = repo.commit('HEAD')
+    for obj in commit.tree.traverse():
+        if any([fnmatch.fnmatchcase(obj.path, pattern) for pattern in patterns]):
+            staged_file = file_io.load_staged_file(obj.data_stream)
+            prev_obj = git_utils.get_file_version(repo, obj.path, 'HEAD~1')
+            prev_staged_file = file_io.load_staged_file(prev_obj.data_stream) if prev_obj else {}
+
+            for param_name, param_metadata in staged_file.items():
+                if not param_name in prev_staged_file or param_metadata != prev_staged_file[param_name]:
+                    oids.add(param_metadata['oid'])
+
+    theta_commits.write_commit_oids(commit.hexsha, oids)
 
 
-def add(args):
+def pre_push(args):
     """
-    Splits model parameters file into parameter groups on filesystem and stages files
+    Pre-push git hook for sending objects to the LFS server
     """
-    logging.debug(f"Running git theta add on {args.file}")
     repo = git_utils.get_git_repo()
-    model_path = git_utils.get_relative_path_from_root(repo, args.file)
-    theta_model_dir = git_utils.get_git_theta_model_dir(repo, model_path, create=True)
+    theta_commits = theta.ThetaCommits(repo)
 
-    checkpoint_handler = checkpoints.get_checkpoint_handler(args.type)
-    logging.debug(f"git theta add using checkpoint type={checkpoint_handler.name}")
-    param_dict = checkpoint_handler.from_file(args.file)
-    
-    for param_dir, param in paths_and_parameters(param_dict, theta_model_dir).items():
-        param_file = os.path.join(param_dir, "params")
-        os.makedirs(param_file, exist_ok=True)
-        file_io.write_tracked_file(param_file, param)
-        git_utils.add_file(param_file, repo)
-
-    for param_file in utils.removed_params(param_dict, utils.walk_parameter_dir(theta_model_dir)):
-        git_utils.remove_file(param_file, repo)
+    gitattributes_file = git_utils.get_gitattributes_file(repo)
+    patterns = git_utils.get_gitattributes_tracked_patterns(gitattributes_file)
 
-    git_utils.add_file(model_path, repo)
+    # Read lines of the form <local ref> <local sha1> <remote ref> <remote sha1> LF
+    lines = sys.stdin.readlines()
+    lines_parsed = [re.match('^(?P<local_ref>[^\s]+) (?P<local_sha1>[^\s]+) (?P<remote_ref>[^\s]+) (?P<remote_sha1>[^\s]+)\n$', l) for l in lines]
```

Comment:
We probably want a tighter regex than space splitting. Something like

`\b[0-9a-f]{40}\b`

I assume this format is something we don't control? like it is from git so we can't convert it into json or anything?

Reply:
Yeah, this can be more precise. The format is from the githooks documentation: https://mirrors.edge.kernel.org/pub/software/scm/git/docs/githooks.html

<local ref> SP <local object name> SP <remote ref> SP <remote object name> LF

Object names we know the length of so we can enforce that.